packages/workshop-app/app/routes/launch-editor.tsx

```diff
@@ -63,36 +64,57 @@ export function LaunchEditor({
 	line,
 	column,
 	children,
+	onUpdate,
 }: {
 	line?: number
 	column?: number
 	children: React.ReactNode
+	onUpdate?: (state: string) => void
 } & (
 	| { file: string; appFile?: never; appName?: never }
-	| { file?: never; appFile: string; appName: string }
+	| { file?: never; appFile: string | string[]; appName: string }
 )) {
 	const fetcher = useFetcher<typeof action>()
+
+	switch (fetcher.state) {
+		case 'submitting': {
+			onUpdate?.(fetcher.state)
+		}
+		case 'loading': {
+			const error = fetcher.data?.status === 'error' ? fetcher.data.error : ''
+			if (error) {
+				showToast(document, {
+					title: 'Launch Editor Error',
+					variant: 'Error',
+					content: error,
+				})
+			}
+			onUpdate?.(fetcher.state)
+		}
+	}
```

Comment:
This should *definitely* be within a `useEffect`.

Reply:
ok.

by the way in TouchedFiles we make spacial effort to make each button a form of LaunchEditor. although it is small list and it fast LaunchEditor run for all the button event after one button click

i prefer making one form in TouchedFiles  that submit to /launch-editor.

what do you think?

Reply:
Each button should be its own form.

Reply:
Why?

Reply:
Because I think that's much easier. To be clear, what I mean is each individual LaunchEditor should render a form with a single submit button, and the Open All button should be rendered within a form as well. The unique bit of the Open All button is that its form will have multiple `<input name="appFile" />` entries.

Reply:
It works, but it's more complicated. Just render it as a form with all the files as hidden inputs and we don't need the on submit handler